README.md
# MMM-RadioDe

[![License](https://img.shields.io/github/license/mashape/apistatus.svg?maxAge=2592000)](https://raw.githubusercontent.com/SteffMUC/MMM-RadioDe/master/LICENSE)
[![Build Status](https://img.shields.io/travis/SteffMUC/MMM-RadioDe.svg?maxAge=2592000)](https://travis-ci.org/SteffMUC/MMM-RadioDe)
[![Known Vulnerabilities](https://snyk.io/test/github/steffmuc/mmm-radiode/badge.svg)](https://snyk.io/test/github/steffmuc/mmm-radiode)

This is an extension for the [MagicMirror](https://github.com/MichMich/MagicMirror). It can show/hide a webradio stream provided by [Radio.de](http://www.radio.de)
You can start and stop this web radio stream by hiding the module. It uses the suspend/ resume method of MagicMirror for this.
Currently the setup is rather simple. Potential extensions for the future might be selection of the radio station via remote control plugin, tuning of volume.
Currently I do use the module to start in the evening the favorite radio channel of my children and to stop it again once they do sleep.

![Magic-Mirror Module MMM-RadioDe screenshot](https://github.com/SteffMUC/MMM-RadioDe/blob/master/screenshot/MMM-RadioDe-Screenshot.png)


## Installation
1. Navigate into your MagicMirror's `modules` folder and execute `git clone https://github.com/SteffMUC/MMM-RadioDe.git`. A new folder will appear navigate into it.
2. Execute `npm install` to install the node dependencies. (Currently not really necessary since there are no dependencies).

## Using the module

To use this module, add it to the modules array in the `config/config.js` file:
````javascript
modules: [
	{
		module: 'MMM-RadioDe',
		config: {
			// See 'Configuration options' for more information.
		}
	}
]
````

## Configuration options

The following properties can be configured:


<table width="100%">
	<!-- why, markdown... -->
	<thead>
		<tr>
			<th>Option</th>
			<th width="100%">Description</th>
		</tr>
	<thead>
	<tbody>
		<tr>
			<td><code>station</code></td>
			<td>Station name of the desired radio station. Default (if nothing specified) is Antenne Bayern Hit for Kids. The station name can be obtained from <li><a href="http://www.radio.de">Radio.de</a></li>
			Enter your desired radio station into the search field for radio stations, start it one time in the browser, you can derive the station name from the html address. E.g. Antenne Bayern Hits for Kids resolves to "antennekids"
		</tr>
		<tr>
			<td><code>api_key</code></td>
			<td>Must be specified. Without api_key the browser player will be displayed but will not start. The api_key can be derived one time in your browser, start any radio station of radio.de, click the embed button, you will see an embed code, copy your api_key from there.
			</td>
		</tr>
		<tr>
			<td><code>height</code></td>
			<td>Optional: Specify the height of your player. Default is 92 (pixel). Please specify only as decimal number without pixel as unit.
			</td>
		</tr>
		<tr>
			<td><code>width</code></td>
			<td>Optional: Specify the width of your player. Default is 360 (pixel). Please specify only as decimal number without pixel as unit.
			</td>
		</tr>
	</tbody>
</table>

## Dependencies
- Currently no dependencies

## Example Configurations
If you use this configuration you need to hide/ unhide this module manually, e.g. with [MMM-Remote-Control](https://github.com/Jopyth/MMM-Remote-Control)
````javascript
{
	module: 'MMM-RadioDe',
	position: 'top_right', // This can be any of the regions.
	config: {
					// See 'Configuration options' for more information.
					station: "antenne",
					// can be obtained in embed section of station, e.g. http://antennekids.radio.de/
					api_key: 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX',
					//optional,
					height:"92", // in pixel
					width:"360", // in pixel
	}
},
````
Alternatively you can use the module [MMM-ModuleScheduler](https://github.com/ianperrin/MMM-ModuleScheduler) to start and stop the radio stream automatically.
````javascript
{
        module: 'MMM-RadioDe',
        position: 'top_right', // This can be any of the regions.
        classes: 'scheduler',
        config: {
                // PLAY RADIO BETWEEN 18:05 and 18:15 EVERY DAY
                module_schedule: {from: '05 18 * * *', to: '15 18 * * *' },
								// See 'Configuration options' for more information.
								station: "antenne",
								// can be obtained in embed section of station, e.g. http://antennekids.radio.de/
								api_key: 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX',
								//optional,
								height:"92", // in pixel
								width:"360", // in pixel
                }       
        },
````

## Daily operation
The module can be controlled via notifications to change the currently playing radio station.
If you have installed [MMM-Remote-Control](https://github.com/Jopyth/MMM-Remote-Control) you can change the radio station by accessing the web interface, e.g.:
````javascript
http://localhost:8080/remote?action=NOTIFICATION&notification=RADIO_STATION&payload={%22title%22:%22antennekids%22,%22type%22:%22notification%22}
````
You need to exchange `localhost` with your magic mirror ip address / hostname and put instead of `antennekids` your desired radio station.

Alternatively you could schedule changes to the playing radio station via	the module [MMM-ModuleScheduler](https://github.com/ianperrin/MMM-ModuleScheduler), e.g.
````javascript
{
				module: 'MMM-ModuleScheduler',
				config: {
						// change radio station to antenne at 21:17 each day
						notification_schedule: {
								notification: 'RADIO_STATION',
								schedule: '17 21 * * *',
								payload: {
										type: "notification",
										title: 'antenne'
								}
						}
				}
		},
````			


CHANGELOG.md
# MMM-RadioDe change log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [0.2.0] - 2017-02-19

### Added
- requiresVersion: "2.1.0",
- option to change playing radio station via notification messages   

### Changed
- Added documentation for daily operation in [README.md](README.md# Daily operation)
- Added check for all modules started, before starting radio player

### Fixed
- Known problems: One user reported that player does not start, potentially fixed by starting player only once all modules started

## [0.1.0] - 2017-02-05
### Initial release of the MMM-RadioDe module.


MMM-RadioDe.js

Module.register("MMM-RadioDe",{

    // due to usage of suspend & resume subclass
    requiresVersion: "2.1.0",


    // Default module config.
    defaults: {
    api_key: false,
    station: 'antennekids',
    autoplay: "true",
    player_color: "61ce42", //does not have any influence
    height:"92", // in pixel
    width:"360", // in pixel
    // volume:"70%" not yet implemented, to be done in later version
    },
    display: true,
    all_started: false,


    // Subclass start method.
    start: function() {
        Log.info("Starting module: " + this.name);
        //Log.info("Setting System Volume to " + this.config.volume);
        //exec("amixer -q sset Master " + this.config.volume, null);
    },

    notificationReceived: function (notification, payload, sender) {
      if(notification === "ALL_MODULES_STARTED"){
        this.all_started = true;
        Log.info("Received notification ALL_MODULES_STARTED setting all_started to  : " + this.all_started);
      }else if(notification === "RADIO_STATION"){
        Log.info("Received RADIO_STATION " + payload.title + "   "+ payload.type);
        this.config.station = payload.title;
        this.updateDom();
      }
    },


    getDom: function() {
      var wrapper = document.createElement("div");
      if (this.display && this.all_started) {

        var script = document.createElement("div");
        script.innerHTML = "<div class=\"ng-app-embedded\"><div style=\"display:none;\" class=\"\" data-playertype=\"web_embedded\" data-playstation=\"antennekids\" data-autoplay=\"false\" data-playercolor=\"" + this.config.player_color + "\" data-apikey=\"" + this.config.api_key + "\" data-iframe=\"true\"></div><iframe src=\"http://radio.de/inc/microsite/index.html?playerWidth="+ this.config.width +"px&amp;playerHeight="+ this.config.height + "px&amp;playerType=web_embedded&amp;partnerLogo=null&amp;partnerUrl=null&amp;partnerBacklink=null&amp;partnerName=null&amp;popoutTitle=null&amp;stations=null&amp;listText=null&amp;listSubtext=null&amp;playerColor=" + this.config.player_color + "&amp;showFooter=null&amp;token=null&amp;playStation=" + this.config.station + "&amp;apikey=" + this.config.api_key + "&amp;popupWidth=null&amp;popupHeight=null&amp;autoplay=" + this.config.autoplay + "&amp;iframe=true\" id=\"radioDeIframe0\" name=\"radioDeIframe0\" frameborder=\"0\" style=\"width: "+ this.config.width +"px; height: "+ this.config.height + "px;\" height=\""+this.config.height+"\" width=\""+this.config.height+"\"></iframe></div>";

        // original script code for testing of 403 error on console
        //script.innerHTML = "<div class=\"ng-app-embedded\"><div style=\"display:none;\" class=\"\" data-playertype=\"web_embedded\" data-playstation=\"antennekids\" data-autoplay=\"false\" data-playercolor=\"61ce42\" data-apikey=\"df04ff67dd3339a6fc19c9b8be164d5b5245ae93\" data-iframe=\"true\"></div><iframe src=\"http://antennekids.radio.de/inc/microsite/index.html?playerWidth=360px&amp;playerHeight=92px&amp;playerType=web_embedded&amp;partnerLogo=null&amp;partnerUrl=null&amp;partnerBacklink=null&amp;partnerName=null&amp;popoutTitle=null&amp;stations=null&amp;listText=null&amp;listSubtext=null&amp;playerColor=61ce42&amp;showFooter=null&amp;token=null&amp;playStation=antennekids&amp;apikey=df04ff67dd3339a6fc19c9b8be164d5b5245ae93&amp;popupWidth=null&amp;popupHeight=null&amp;autoplay=true&amp;iframe=true\" id=\"radioDeIframe0\" name=\"radioDeIframe0\" frameborder=\"0\" style=\"width: 360px; height: 92px;\" height=\"92\" width=\"360\"></iframe></div>";


        wrapper.appendChild(script);
        }

      return wrapper;
  	  },

      suspend: function(){
        this.display = false;
        this.updateDom();
      },

      resume: function(){
        this.display = true;
        this.updateDom();
      },



});
